test/unittest/ext/llm/test_model.py

```diff
@@ -0,0 +1,112 @@
+    import peft
+except ImportError:
+    bitsandbytes = None
+    peft = None
+
+
+@pytest.mark.parametrize(
+    "db", [DBConfig.mongodb_empty, DBConfig.sqldb_empty], indirect=True
+)
+def test_predict(db):
+    """Test chat."""
+    model = LLM(identifier="llm", model_name_or_path=TEST_MODEL_NAME)
+
+    check_predict(db, model)
+
+
+@pytest.mark.parametrize(
+    "db", [DBConfig.mongodb_empty, DBConfig.sqldb_empty], indirect=True
+)
+def test_model_as_listener_model(db):
+    model = LLM(identifier="llm", model_name_or_path=TEST_MODEL_NAME)
+    check_llm_as_listener_model(db, model)
+
+
+@pytest.mark.skipif(
+    not bitsandbytes or not peft, reason="The peft and bitsandbytes are not installed"
+)
+@pytest.mark.parametrize("db", [DBConfig.mongodb_empty], indirect=True)
+def test_training(db, tmpdir):
+    datas = []
+    for i in range(32 + 8):
+        text = f"{i}+1={i+1}"
+        fold = "train" if i < 32 else "valid"
+        datas.append({"text": text, "id": str(i), "fold": fold})
+
+    db.execute(Collection("doc").insert_many(list(map(Document, datas))))
+    select = Collection("doc").find()
+
+    model = LLM(
+        identifier="llm",
+        model_name_or_path="facebook/opt-125m",
+        tokenizer_kwags=dict(model_max_length=64),
+    )
+    training_configuration = LLMTrainingConfiguration(
+        identifier="llm-finetune",
+        output_dir=tmpdir,
+        lora_r=64,
+        lora_alpha=16,
+        lora_dropout=0.05,
+        num_train_epochs=3,
+        per_device_train_batch_size=1,
+        per_device_eval_batch_size=1,
+        gradient_accumulation_steps=8,
+        evaluation_strategy="steps",
+        eval_steps=1,
+        save_strategy="steps",
+        save_steps=1,
+        save_total_limit=3,
+        learning_rate=2e-5,
+        weight_decay=0.0,
+        warmup_ratio=0.03,
+        lr_scheduler_type="cosine",
+        logging_strategy="steps",
+        logging_steps=1,
+        gradient_checkpointing=True,
+        report_to=[],
+        max_length=64,
+    )
+
+    model.fit(
+        X="text",
+        db=db,
+        select=select,
+        configuration=training_configuration,
+        prefetch_size=1000,
+    )
+
+    checkpoints = os.listdir(tmpdir)
+    checkpoints = [
+        checkpoint for checkpoint in checkpoints if checkpoint.startswith("checkpoint")
+    ]
+    checkpoints = sorted(checkpoints, key=lambda x: int(x.split("-")[-1]))
+
+    # Test checkpoints
+    assert len(checkpoints) == 3
+
+    # Test multi-lora adapters
+    model = LLM(
+        identifier="llm-inference",
+        model_name_or_path="facebook/opt-125m",
+        tokenizer_kwags=dict(model_max_length=64),
+    )
+    for checkpoint in checkpoints:
+        model.add_adapter(os.path.join(tmpdir, checkpoint), adapter_name=checkpoint)
+
+    db.add(model)
+    db.predict(model.identifier, "1+1=")
+    for checkpoint in checkpoints:
+        db.predict(model.identifier, "1+1=", adapter_name=checkpoint)
```

Comment:
I have some ideas around this. It's possible we can solve adapters without adding this extra parameter.

Reply:
To be discussed tomorrow.

Reply:
How about this:
```python
@dc.dataclass
class Lora(Model):
    identifier: str = ""
    lora_path: str = "facebook/opt-125m"
    base_model_identifier: t.Optional[str] = None

    # below are the same as LLM for define how to load the base model
    model_name_or_path: str = "facebook/opt-125m"
    bits: t.Optional[int] = None
    model_kwargs: t.Dict = dc.field(default_factory=dict)
    tokenizer_kwags: t.Dict = dc.field(default_factory=dict)
    prompt_template: str = "{input}"
    prompt_func: t.Optional[t.Callable] = dc.field(default=None)

    type_id = "lora"

    def pre_create(self, db: Datalayer):
        # try to load base model, if not exist, create a new one
        base_model = db.load("model", self.base_model_identifier)
        if base_model is None:
            base_model = LLM(
                identifier=self.base_model_identifier,
                model_kwargs=self.model_kwargs,
                tokenizer_kwags=self.tokenizer_kwags,
                prompt_template=self.prompt_template,
                prompt_func=self.prompt_func,
            )

        # add the base model to db for sharing 
        db.add(base_model)
        base_model.add_adapter(self.lora_path, self.identifier)

        self.base_model = base_model
        super().pre_create(db)

    def _predict(self, *args, **kwargs):
        # all the requrests will be sent to the base model
        kwargs["adapter_name"] = self.identifier
        return self.base_model._predict(*args, **kwargs)

# load lora model based on a exist base model
llm = LLM(identifier='llm', model_name_or_path='facebook/opt-125m')

db = ....
db.add(Lora('lora1', lora_path='checkpoint_path_1', base_model_identifier='llm'))
db.predict('lora1', ...)


# load lora model directly
db = ....
db.add(Lora('lora1', lora_path='checkpoint_path_1', model_name_or_path='facebook/opt-125m', bits=4))
db.predict('lora1', ...)

db.add(Lora('lora2', lora_path='checkpoint_path_2', model_name_or_path='facebook/opt-125m', bits=4))
db.predict('lora2', ...)

```

Reply:
@blythed Please help to take a look at this design when you have time, thanks

Reply:
The current usage method is as follows. These models will share the underlying model. If the initialization parameters of the underlying model are the same

```python
from superduperdb import superduper
db = superduper("mongomock://llm-finetune")
from superduperdb.ext.llm import LLM
llm1 = LLM(identifier='adapter1', model_name_or_path="facebook/opt-125m" , prompt_template="1+{input}=", adapter_id='outputs/llm-finetune/checkpoint-10/')
llm2 = LLM(identifier='adapter2', model_name_or_path="facebook/opt-125m" , prompt_template="1+{input}=", adapter_id='outputs/llm-finetune/checkpoint-20/')
db.add(llm1)
db.add(llm2)
db.predict('adapter1', "2")
db.predict('adapter2', "2")
```